File: ocaml/functions/list_operations.ml
(* List operations in OCaml *)

let demo () =
  print_endline "\n=== List Operations in OCaml ===\n";

  let numbers = [1; 2; 3; 4; 5] in
  Printf.printf "Numbers: [%s]\n" (String.concat "; " (List.map string_of_int numbers));

  (* List length *)
  Printf.printf "Length: %d\n" (List.length numbers);

  (* Head and tail *)
  Printf.printf "Head: %d\n" (List.hd numbers);
  Printf.printf "Tail: [%s]\n" (String.concat "; " (List.map string_of_int (List.tl numbers)));

  (* Prepend *)
  let prepended = 0 :: numbers in
  Printf.printf "\nPrepend 0: [%s]\n" (String.concat "; " (List.map string_of_int prepended));

  (* Append *)
  let appended = numbers @ [6; 7; 8] in
  Printf.printf "Append: [%s]\n" (String.concat "; " (List.map string_of_int appended));

  (* Map *)
  let squared = List.map (fun x -> x * x) numbers in
  Printf.printf "\nMap (square): [%s]\n" (String.concat "; " (List.map string_of_int squared));

  (* Filter *)
  let evens = List.filter (fun x -> x mod 2 = 0) numbers in
  Printf.printf "Filter (even): [%s]\n" (String.concat "; " (List.map string_of_int evens));

  (* Fold *)
  let sum = List.fold_left (+) 0 numbers in
  Printf.printf "\nFold (sum): %d\n" sum;

  (* Reverse *)
  let reversed = List.rev numbers in
  Printf.printf "Reversed: [%s]\n" (String.concat "; " (List.map string_of_int reversed))

let () = demo ()

File: prolog/functions/list_operations.pl
% List operations in Prolog

% List predicates
demo :-
    writeln('\n=== List Operations in Prolog ===\n'),

    % Create list
    Numbers = [1, 2, 3, 4, 5],
    format('Numbers: ~w~n', [Numbers]),

    % List length
    length(Numbers, Len),
    format('Length: ~w~n', [Len]),

    % Head and tail
    [H|T] = Numbers,
    format('\nHead: ~w~n', [H]),
    format('Tail: ~w~n', [T]),

    % Append
    append(Numbers, [6, 7, 8], Appended),
    format('\nAppend: ~w~n', [Appended]),

    % Member
    (member(3, Numbers) -> writeln('Member 3: true') ; writeln('Member 3: false')),
    (member(10, Numbers) -> writeln('Member 10: true') ; writeln('Member 10: false')),

    % Reverse
    reverse(Numbers, Reversed),
    format('\nReversed: ~w~n', [Reversed]),

    % Sort
    Unsorted = [5, 2, 8, 1, 9, 3],
    sort(Unsorted, Sorted),
    format('\nSorted: ~w~n', [Sorted]),

    % Sum (using fold)
    sumlist(Numbers, Sum),
    format('Sum: ~w~n', [Sum]),

    % Max element
    max_list(Numbers, Max),
    format('Max: ~w~n', [Max]),

    % Min element
    min_list(Numbers, Min),
    format('Min: ~w~n', [Min]),

    % Take first n elements
    length(Prefix, 3),
    append(Prefix, _, Numbers),
    format('\nFirst 3: ~w~n', [Prefix]),

    % Map (using maplist)
    maplist(square, Numbers, Squared),
    format('\nSquared: ~w~n', [Squared]),

    % Filter even numbers
    include(even_number, Numbers, Evens),
    format('Evens: ~w~n', [Evens]),

    halt.

% Helper predicates
square(X, Y) :- Y is X * X.

even_number(X) :- 0 is X mod 2.

% Run demo
:- initialization(demo).
